src/tensor/cudense.jl

```diff
@@ -117,11 +118,52 @@ function permutedims!!(B::CuDenseTensor{ElT,N},
                        A::CuDenseTensor{ElT,0},
                        perm::NTuple{N,Int},
                        f=(r,t)->permute!(r,t)) where {N, ElT<:Number}
-    Cis = permute(inds(B), perm)
+    Cis = ITensors.NDTensors.permute(inds(B), perm)
     Cs = f(B, A)
     return Tensor(Dense(vec(Cs)), Cis) 
 end
 
+function _contract_scalar!(R::CuDenseTensor{ElR}, labelsR,
+                           T1::Number, labelsT1,
+                           T2::Number, labelsT2,
+                           α = one(ElR), β = zero(ElR)) where {ElR}
+  if iszero(β)
+    copyto!(data(R), [α * T1 * T2])
+  elseif iszero(α)
+    copyto!(data(R), β.*data(R))
+  else
+    copyto!(data(R), [α * T1 * T2] .+ β.*data(R))
+  end
+  return R
+end
+
+function _contract_scalar!(R::CuDenseTensor{ElR,NR}, labelsR,
+                           T₁::CuDenseTensor, labelsT₁,
+                           T₂::CuDenseTensor, labelsT₂,
+                           α = one(ElR), β=zero(ElR)) where {ElR,NR}
+    if nnz(T₁) == nnz(T₂) == 1
+        new_R = Tensor(Dense(data(store(T₁)).*data(store(T₂))), inds(R))
+        copyto!(store(R), store(new_R))
+    elseif nnz(T₁) == 1
```

Comment:
Hi Katie,

I added this branch of the code: https://github.com/ITensor/NDTensors.jl/blob/v0.1.29/src/dense.jl#L788 because I was finding that for contractions that involved scalar-like tensors it was a big speedup to not go through the `compute_contraction_properties` code logic and directly do permutations or call BLAS level 1 functions. Probably you already know and skipped that for now since the logic is a bit convoluted. It mostly was relevant for block sparse calculations with many blocks of dimension 1. Probably a lot of the logic for `_contract_scalar_maybe_perm!` could be handled directly by a single `cuTENSOR` function call. Just a heads up to give you context on the new changes.

For my own understanding, I thought at this point cuTENSOR was being used for the contractions, does the code ever reach this point? Couldn't it overload at a higher level, like the higher level `contract!` function [here](https://github.com/ITensor/NDTensors.jl/blob/v0.1.29/src/dense.jl#L793-L796) so it never reaches the messy `_contract_scalar!` and `compute_contraction_properties!` code?

-Matt

Reply:
In fact in the call [here](https://github.com/ITensor/ITensorsGPU.jl/pull/8/files#diff-cdacba6cce2c0bc71d7b13970d661a90caeb29f92579cf2aa39f3b1730ae625aR167-R171) it seems like it never makes use of the `props::ContractionProperties` input (which is only used to determine how to turn the contraction into a matrix-matrix multiplication with permutations) so it seems like it could all be overloaded at a higher level to simplify the logic a bit.